### DISCUSSION TITLE
Issue with conditional require of field if array contains a const

### DISCUSSION THREAD

**Discussion by andlbrei** [json-schema-org/community#822] [Q&A] [answered]
Hello!
I'm scratching my head a bit here, wondering why this isn't working.
Hopefully some of you can assist me in this issue, which I feel like I should have figured out by now...

This is a my schema
```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "type": "object",
  "properties": {
    "otherIncomeSource": {
      "type": "array",
      "items": {
        "type": "string"
      }
    },
    "otherIncomeSourceOtherComment": {
      "type": "string"
    }
  },
  "if": {
    "properties": {
      "otherIncomeSource": {
        "type": "array",
        "contains": {
          "const": "OTHER"
        }
      }
    }
  },
  "then": {
    "required": [
      "otherIncomeSourceOtherComment"
    ]
  }
}
```

When validating against an empty object I get an error:
```
Message:
    Required properties are missing from object: otherIncomeSourceOtherComment.
Schema path:
    #/then/required
```

If I set `otherIncomeSource` as required, and include it in the data as an empty array, then the `if/else` validates as expected.

What am I doing wrong?

I want `otherIncomeSource` to be optional, and if it is present and it has an item with the value `OTHER`, then I want the field `otherIncomeSourceOtherComment` to be required.

You can find an example here:
https://www.jsonschemavalidator.net/s/bx5E08xw

**Comment by jviotti** [comment, marked as answer]
Hey @andlbrei ,

The problem is on your `if`. Your check there is essentially saying that "if (if `otherIncomeSource ` is defined and matches the given schema...)", but doesn't enforce `otherIncomeSource` to be present. Hence, an empty schema passes the `if` and you get into `then`, which forces `otherIncomeSourceOtherComment` to be required. Instead, you probably want to add `required` to your `if` as well, so that the empty instance doesn't pass:

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "type": "object",
  "properties": {
    "otherIncomeSource": {
      "type": "array",
      "items": {
        "type": "string"
      }
    },
    "otherIncomeSourceOtherComment": {
      "type": "string"
    }
  },
  "if": {
    "required": [ "otherIncomeSource" ],
    "properties": {
      "otherIncomeSource": {
        "type": "array",
        "contains": {
          "const": "OTHER"
        }
      }
    }
  },
  "then": {
    "required": [
      "otherIncomeSourceOtherComment"
    ]
  }
}
```

Schemas can quickly get complicated. If it helps, what I typically do is use https://github.com/sourcemeta/jsonschema/blob/main/docs/test.markdown to write little unit tests for all the schemas I write.

**Reply by andlbrei** [reply]
Ah, great. I knew it would be simple... Thank you!
